qtpyvcp/widgets/virtual_input/virtual_input.py

```diff
@@ -0,0 +1,140 @@
+
+    @Slot()
+    def on_enter_key_clicked(self):
+        self.send_key_(Qt.Key_Enter)
+
+    @Slot()
+    def on_esc_key_clicked(self):
+        self.send_key_(Qt.Key_Escape)
+        self.hide()
+
+    @Slot()
+    def on_backspace_key_clicked(self):
+        self.send_key_(Qt.Key_Backspace)
+
+    @Slot()
+    def on_left_key_clicked(self):
+        self.send_key_(Qt.Key_Left)
+
+    @Slot()
+    def on_right_key_clicked(self):
+        self.send_key_(Qt.Key_Right)
+
+    @Slot()
+    def on_up_key_clicked(self):
+        self.send_key_(Qt.Key_Up)
+
+    @Slot()
+    def on_down_key_clicked(self):
+        self.send_key_(Qt.Key_Down)
+
+    @Slot()
+    def on_tab_key_clicked(self):
+        self.send_key_(Qt.Key_Tab)
+
+    @Slot()
+    def on_home_key_clicked(self):
+        self.send_key_(Qt.Key_Home)
+
+    @Slot()
+    def on_end_key_clicked(self):
+        self.send_key_(Qt.Key_End)
+
+    @Slot()
+    def on_page_up_key_clicked(self):
+        self.send_key_(Qt.Key_PageUp)
+
+    @Slot()
+    def on_page_down_key_clicked(self):
+        self.send_key_(Qt.Key_PageDown)
+
+    @Slot()
+    def on_caps_key_clicked(self):
+        self.caps_on = not self.caps_on
+        for btn in self.buttonGroup.buttons():
+            if self.caps_on:
+                btn.setText(btn.text().upper())
+            else:
+                btn.setText(btn.text().lower())
+
+    def set_focus_object(self, value):
+        self.focus_object = value
+        self.position_keyboard_to_object(self.focus_object)
+
+    def position_keyboard_to_object(self, obj):
+        screen = QtWidgets.QDesktopWidget().screenGeometry()
+        pos = obj.mapToGlobal(QPoint(0, obj.height()))
+
+        if pos.y() + self.height() > screen.height():
+            pos.setY(pos.y() - self.height() - obj.height())
+        if pos.x() + self.width() > screen.width():
+            pos.setX(pos.x() - self.width() + obj.width())
+        if pos.x() + self.width() > screen.width():
+            pos.setX(pos.x() - (pos.x() + self.width() - screen.width()))
+
+        self.move(pos.x(), pos.y())
+
+    def send_key_(self, key):
+        event = QKeyEvent(QEvent.KeyPress, key, Qt.NoModifier)
+        QGuiApplication.postEvent(self.focus_object, event)
+
+        event = QKeyEvent(QEvent.KeyRelease, key, Qt.NoModifier)
+        QGuiApplication.postEvent(self.focus_object, event)
+
+    @staticmethod
+    def try_enable_virtual_input(app):
+        ini = linuxcnc.ini(os.getenv("INI_FILE_NAME"))
+        if ini.find("DISPLAY", "TOUCHSCREEN") not in ['yes', '1', 'True']:
+            return
+
+        np = VirtualInput('numpad.ui')
+        kb = VirtualInput('keyboard.ui')
+
+        def show_keyboard(_, new):
+            kb.hide()
+            np.hide()
+
+            if isinstance(new, QAbstractSpinBox):
+                np.set_focus_object(new)
+                np.show()
+            elif hasattr(new, 'is_read_only') and hasattr(new, 'text') and not new.is_read_only():
```

Comment:
This is workaround for the GCodeEditor. The build was failing trying to find QScintilla and I could not get
it to install. Rather than fight it I think longer term it might be good to have input controls inherit from another class; InputMetaData or something like that. This class can have methods, show_numpad, show_keyboard, and then we don't have to play the guessing game on what type of input to show, if any. Adding InputNumber and InputText classes can override the show_numpad, show_keyboard methods accordingly. This would mean that inheriting from InputNumber, or InputText would be enough to get proper input support from the numpad or keybaord.